gherkin/php/src/TokenMatcher.php

```diff
@@ -211,12 +221,41 @@ public function match_Language(Token $token): bool
 
             $this->currentDialect = $this->dialectProvider->getDialect($language, $token->getLocation());
 
+            $this->updateKeywordTypeMappings($this->currentDialect);
+
             return true;
         }
 
         return false;
     }
 
+    private function updateKeywordTypeMappings(GherkinDialect $dialect): void
+    {
+        $this->keywordTypes = [];
+        $this->addKeywordTypeMappings($dialect->getGivenKeywords(), KeywordType::CONTEXT);
+        $this->addKeywordTypeMappings($dialect->getWhenKeywords(), KeywordType::ACTION);
+        $this->addKeywordTypeMappings($dialect->getThenKeywords(), KeywordType::OUTCOME);
+        $this->addKeywordTypeMappings($dialect->getAndKeywords(), KeywordType::CONJUNCTION);
+        $this->addKeywordTypeMappings($dialect->getButKeywords(), KeywordType::CONJUNCTION);
+    }
+
+    /** @param array<string> $keywords */
+    private function addKeywordTypeMappings(array $keywords, KeywordType $type): void
+    {
+        foreach ($keywords as $keyword) {
+            $this->keywordTypes[$keyword][] = $type;
+        }
+    }
+
+    private function getKeywordType(string $keyword): KeywordType
+    {
+        if (!array_key_exists($keyword, $this->keywordTypes) || count($this->keywordTypes[$keyword]) !== 1) {
+            return KeywordType::UNKNOWN;
+        }
+
+        return $this->keywordTypes[$keyword][0];
```

Comment:
I've added a comment to avoid any ambiguity

Reply:
Actually the PHP version we support is high enough we can use https://www.php.net/manual/en/function.array-key-first.php

Reply:
Sigh ignore me we can use `current()` rather than `[0]` to get the 'next' value irrespective of key

Reply:
I must say that I prefer `[0]` as it is explicit we want the first item in the array. And anyone, even junior developer, should be comfortable with that notation.

`current` implies an iteration, which there is not here. And using `current` without `reset` to get the first item does not looks safe enough imo

Reply:
Yeah my point earlier was that without it being defined as a `list` we had no guarantees that `0` is the correct key

i.e. `['foo'=>'bar']` has `count(...) === 1`

Reply:
However I have just pushed a commit which removes the `array_key_exists` before the `count` on the array, as `count` will properly returns `0`

Reply:
If the key doesn't exist there will be a warning (because accessing an unset key) and a fatal error (because then trying to count `null`)

https://3v4l.org/HUONa

Reply:
Here's another option:

```php
count($this->keywordTypes[$keyword] ?? [])==1
```

Reply:
> Yeah my point earlier was that without it being defined as a `list` we had no guarantees that `0` is the correct key
> 
> i.e. `['foo'=>'bar']` has `count(...) === 1`

I have updated to type annotation of `$this->keywordTypes` to make it an explicit `list` actually



> If the key doesn't exist there will be a warning (because accessing an unset key) and a fatal error (because then trying to count `null`)
> 
> https://3v4l.org/HUONa

My bad. I have misread the doc, and found a bad online php interpreter :'(
I revert this right now

Reply:
> Here's another option:
> 
> ```
> count($this->keywordTypes[$keyword] ?? [])==1
> ```

Nice one indeed 👍

I suggest the following to make it absolutely clear what we actually want to achieve:
```php
    private function getKeywordType(string $keyword): KeywordType
    {
        if (!array_key_exists($keyword, $this->keywordTypes)) {
            return KeywordType::UNKNOWN;
        }

        // We want $this->keywordTypes[$keyword] having **only 1 entry**
        if (count($this->keywordTypes[$keyword]) !== 1) {
            return KeywordType::UNKNOWN;
        }

        return $this->keywordTypes[$keyword][0];
    }
``` 

This can be less performant and redundant, but it removes any doubt on what the code is actually doing